test/functional/test-xhr.js

```diff
@@ -217,6 +223,63 @@ describe('XHR', function() {
 
     describe(test.desc, () => {
 
+      describe('assertSuccess', () => {
+        function createResponseInstance(body, init) {
+          if (test.desc == 'Native') {
+            return new Response(body, init);
+          } else {
+            init.responseText = body;
+            return new FetchResponse(init);
+          }
+        }
+        const mockXhr = {
+          status: 200,
+          headers: {
+            'Content-Type': 'plain/text',
+          },
+          getResponseHeader: () => '',
+        };
+
+        it('should resolve if success', () => {
+          mockXhr.status = 200;
+          return assertSuccess(createResponseInstance('', mockXhr))
+              .then(response => {
+                expect(response.status).to.equal(200);
+              }).should.not.be.rejected;
+        });
+
+        it('should reject if error', () => {
+          mockXhr.status = 500;
+          return assertSuccess(createResponseInstance('', mockXhr))
+              .then(response => {
+                expect(response.status).to.equal(500);
+              }).should.be.rejectedWith(/HTTP error 500/);
+        });
+
+        it('should parse json content when error', () => {
+          mockXhr.status = 500;
+          mockXhr.responseText = '{"a": "hello"}';
+          mockXhr.headers['Content-Type'] = 'application/json';
+          mockXhr.getResponseHeader = () => 'application/json';
+          return assertSuccess(createResponseInstance('{"a": 2}', mockXhr))
+              .catch(error => {
+                expect(error.responseJson).to.be.defined;
+                expect(error.responseJson.a).to.equal(2);
+              });
+        });
+
+        it('should not return a resolved indented promise', () => {
+          mockXhr.status = 500;
+          mockXhr.responseText = '{"a": "hello"}';
+          mockXhr.headers['Content-Type'] = 'application/json';
+          mockXhr.getResponseHeader = () => 'application/json';
+          const promise = assertSuccess(
+              createResponseInstance('{"a": 2}', mockXhr));
+          promise.should.be.rejected;
```

Comment:
@dvoytenko this catches the problem with returning the resolved promise. I don't understand it well tbh, but my guess is that the promise is resolving and the `.should.be.rejected` won't catch this if we returned the resolved promise.